### DISCUSSION TITLE
Am I missing something with mocks?

### DISCUSSION THREAD

**Discussion by KaloyanYosifov** [vitest-dev/vitest#433] [Q&A] [answered]
Hey guys,

I am trying to test one of my classes, but it seems I cannot mock it for some reason. I followed the guide to mock a module, but it doesn't seem to work for me.

Here is the code I have written below, I've tried it without an alias, with an alias, without the extension, and with it and it doesn't seem to mock. Therefore I decided to check the source code and I see that the mock function is empty https://github.com/vitest-dev/vitest/blob/main/packages/vitest/src/integrations/vi.ts#L81. 

Correct me If I am wrong, but is the mock function bound somewhere else? I added a console log to it and it printed, so my guess is that mocking modules is not yet implemented, yet people are mocking modules. 

```
/**
 * External dependencies.
 */
import { describe, it, vi } from 'vitest';

/**
 * Internal dependencies.
 */
import { loadValidations } from '@js/helpers/bycontract.js';
import CreateProposalAction from '@js/actions/CreateProposalAction.js';

vi.mock('@js/actions/CreateProposalAction.js', async () => {
  console.log('e');

  return {
    execute: () => {
      console.log('e');
    },
  };
});

loadValidations();

describe('Create Proposal Action', () => {
  it('hacks', async () => {
    const action = new CreateProposalAction();
    await action.execute({
      title: 'Hello',
      description: 'sup',
      targets: ['0x0000000000000'],
      callData: ['0x0000000000000'],
    });
  });
});
```

**Comment by sheremet-va** [comment, marked as answer]
You are mocking a class that is stored in `default`, so you should return something like this from factory:

```ts
vi.mock('@js/actions/CreateProposalAction.js', () => {
  return {
     default: class {
        execute() { 
          console.log('') 
        }
     }
  }
})
```

If you mock like in your example, your module would look like this:

```ts
export const execute = () => console.log('123')
```

> Correct me If I am wrong, but is the mock function bound somewhere else? 

Mock functions inside `VitestUtils` are just type hints, they are overwritten in compile time.

----

If you don't want to mock entire module, you can mock only method, using `spyOn`:

```ts
    const action = new CreateProposalAction();

    vi.spyOn(action, 'execute')

    await action.execute({
      title: 'Hello',
      description: 'sup',
      targets: ['0x0000000000000'],
      callData: ['0x0000000000000'],
    });

    expect(action.execute).toHaveBeenCalled()
```

You can also try mocking only one method for the whole suite:

```ts
vi.mock('@js/actions/CreateProposalAction.js', async () => {
  const { default: CreateProposalAction } = await vi.importActual('@js/actions/CreateProposalAction.js')
  vi.spyOn(CreateProposalAction, 'execute')
  // or vi.spyOn(CreateProposalAction.prototype, 'execute') - can't check right now
  return { default: CreateProposalAction }
})
```

**Reply by KaloyanYosifov** [reply]
Just tried it with

```
vi.mock('@js/actions/CreateProposalAction.js', () => {
  return {
    default: class {
      execute() {
        console.log('');
      }
    },
  };
});
```

and it didn't work. Used the relative path and still, it didn't work.

Here is my vite config if it will help:

```
/**
 * External dependencies.
 */
import { defineConfig } from 'vite';
import vue from '@vitejs/plugin-vue';

/**
 * Internal dependencies.
 */
import webpackConfig from './webpack.config';

// https://vitejs.dev/config/
export default defineConfig({
  envPrefix: 'TMP_',
  resolve: {
    alias: webpackConfig.resolve.alias,
  },
  plugins: [vue()],
  optimizeDeps: {
    exclude: ['vue-lottie'],
  },
  test: {},
});
```

and webpack.config.js

```
/**
 * External dependencies.
 */
const path = require('path');

/**
 * Internal dependencies.
 */

/**
 * !!!!!!!!!!!!!!!
 *
 * This config is only used, so we have aliasing autocomplete on IDE
 *
 * !!!!!!!!!!!!!!!
 */

module.exports = {
  resolve: {
    alias: {
      util: 'util',
      zlib: 'browserify-zlib',
      process: 'process/browser',
      stream: 'stream-browserify',
      '@': path.resolve(__dirname, './src'),
      '@js': path.resolve(__dirname, './src/js'),
      '@scss': path.resolve(__dirname, './src/scss'),
      '@contracts': path.resolve(__dirname, './contracts'),
    },
  },
};
```

**Reply by sheremet-va** [reply]
I need more information to help. Please, give reproduction. Preferably using https://stackblitz.com/

**Reply by KaloyanYosifov** [reply]
Reproduced the issue here -> https://stackblitz.com/edit/vitejs-vite-tmqc9d?file=tests/actions/CreateProposalAction.spec.js

There is a script in package.json called "test", so you can run `npm run test` or `vitest run` whichever you prefer.

**Reply by sheremet-va** [reply]
It was a bug in vitest. Should be fixed in https://github.com/vitest-dev/vitest/commit/f9fabaa72364bf04c39e967ca95294e2d4452224

Fix will be in the next release

**Reply by KaloyanYosifov** [reply]
I see. Thank you very much @sheremet-va 🙏 

Have a great day!

**Reply by sheremet-va** [reply]
> I see. Thank you very much @sheremet-va 🙏 
> 
> 
> 
> Have a great day! 

Can you confirm it works in latest version?

**Reply by KaloyanYosifov** [reply]
Yes, it is working! 

Thank you 🙏
